[pang-yy] iP

pang-yy opened the pull request:
# Fido

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Fido frees your mind of having to remember things you need to do. It's,
- text-based
- easy to use
- super fast to use

All you need to do is,
1. download it from [here](https://github.com/pang-yy/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    try {
        new Fido("data/task.txt").run();
    } catch (IOException e) {
        System.out.println(e.getMessage());
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

pang-yy committed:
Rename Duke.java to Fido.java

pang-yy committed:
Update Duke to Fido

pang-yy committed:
Update greet messages

pang-yy committed:
Add echo function

pang-yy committed:
Add ability to store user input and display back

Add ability to store whatever text entered by the user and display them back to the user when requested

pang-yy committed:
Add ability to mark

Add the ability to mark tasks as done and change the status back to not done.

pang-yy committed:
Add support for tracking three types of tasks

Add support for tracking todo, deadline or event tasks

pang-yy committed:
Add test case for text ui

pang-yy committed:
Remove the use of Optional

pang-yy committed:
Add error handling

pang-yy committed:
Add support for deleting tasks from the list

pang-yy committed:
Change protected field in Task.java to private

pang-yy committed:
Change Task class to abstract class

pang-yy committed:
Update greet message and help menu

pang-yy committed:
Minor changes

pang-yy committed:
Combine mark, unmark and delete in switch case

pang-yy committed:
Update Fido ascii art

pang-yy committed:
Remove extra map

pang-yy committed:
Implement save task to disk

pang-yy committed:
Merge branch 'branch-Level-7'

pang-yy committed:
Remove data file

pang-yy committed:
Implement basic date and time functionality

pang-yy committed:
Fix data file not cleared before writting to it

pang-yy committed:
Add due command to view list of due tasks

pang-yy committed:
Merge branch 'branch-Level-8'

pang-yy committed:
Refactor the code to extract out closely related code as classes

pang-yy committed:
Divide classes into packages

pang-yy committed:
Merge branch 'add-gradle-support'

pang-yy committed:
Integrate Gradle into Fido

pang-yy committed:
Rename UI to Ui

pang-yy committed:
Add JUnit tests to test the behavior of the code

pang-yy committed:
Add JavaDoc comments to the code

pang-yy committed:
Reorder import statements

pang-yy committed:
Rename UI.java to Ui.java

pang-yy committed:
Rename UI.java to Ui.java

pang-yy committed:
Tweak the code to comply with a coding standard

pang-yy committed:
Rename UI.java to Ui.java

pang-yy committed:
Implement find tasks

pang-yy committed:
Merge branch 'branch-A-CodingStandard'

pang-yy committed:
Merge branch 'branch-Level-9'

pang-yy committed:
Merge branch 'branch-Level-9'

pang-yy committed:
Merge branch 'master' of https://github.com/pang-yy/ip

pang-yy committed:
Add checkStyle to detect coding style violations

pang-yy committed:
Modify code to pass some checkstyle

pang-yy committed:
Merge branch 'branch-A-CheckStyle'

pang-yy committed:
Add a simple GUI to the chatbot

pang-yy committed:
Merge branch 'branch-Level-10'

pang-yy committed:
Sends greet message on GUI start up

DanielJames0302 commented on src/main/java/fido/task/Deadline.java:
```diff
@@ -0,0 +1,143 @@
+     *         {@code false} otherwise.
+     */
+    @Override
+    public boolean isDue() {
```
Good boolean method naming

DanielJames0302 commented on src/main/java/fido/task/Deadline.java:
```diff
@@ -0,0 +1,143 @@
+     */
+    @Override
+    public String fileFormat() {
+        return String.format("D%s%s%s%s%s%s",
```
I like how you organize the output string using `String.format` to match the pattern

DanielJames0302 commented on src/main/java/fido/task/Event.java:
```diff
@@ -0,0 +1,156 @@
+     * @param name The name or description of the task.
+     * @param startDate The start date of the task as a {@code String} in "MMM dd yyyy" format.
+     * @param endDate The start date of the task as a {@code String} in "MMM dd yyyy" format.
+     */
```
I like your clear and concise javadoc in this place and other places as well

DanielJames0302 commented on src/main/java/fido/task/Event.java:
```diff
@@ -0,0 +1,156 @@
+    public static Event of(String rawInput) throws FidoException {
+        String[] ins = rawInput.split("(/from|/to)");
+        if (ins.length == 0) {
+            throw new FidoException(FidoException.ErrorType.EVENT_EMPTY_DESCRIPTION);
```
Good use of enums for the exception

DanielJames0302 reviewed (commented):
LGTM. Well done.

zyonwee commented on src/main/java/fido/exception/FidoException.java:
```diff
@@ -0,0 +1,76 @@
+    /**
+     * Represents the various types of errors that can occur in the Fido chatbot.
+     */
+    public enum ErrorType {
+        NOT_VALID_INDEX,
+        NOT_VALID_NUMBER,
+        EMPTY_DESCRIPTION,
+        DEADLINE_EMPTY_DESCRIPTION,
+        DEADLINE_EMPTY_DATE,
+        EVENT_EMPTY_DESCRIPTION,
+        EVENT_EMPTY_DATE,
+        EVENT_MISSING_DATE,
+        NOT_VALID_DATE,
+
+        NOT_VALID_FILEPATH,
+        FILE_READ_ERROR,
+        FILE_WRITE_ERROR,
+    }
```
I like your use of Enums

zyonwee commented on src/main/java/fido/storage/Parser.java:
```diff
@@ -0,0 +1,57 @@
+     * @param content The content read from file.
+     * @return List of {@code Task} objects.
+     */
+    public static List<Task> parseFromFile(String content) throws FidoException {
+        if (content.isEmpty()) {
+            return List.<Task>of();
+        }
+        return Arrays.asList(content.split("\n"))
+                .stream()
+                .map(line -> line.split(Parser.DIVIDER))
+                .map(arr -> arr[0].equals("T")
+                        ? (arr[1].equals("false") ? (new Todo(arr[2])) : (new Todo(arr[2]).mark()))
+                        : (arr[0].equals("D")
+                                ? (arr[1].equals("false") ? (new Deadline(arr[2], arr[3]))
+                                        : (new Deadline(arr[2], arr[3]).mark()))
+                                : (arr[1].equals("false") ? (new Event(arr[2], arr[3], arr[4]))
+                                        : (new Event(arr[2], arr[3], arr[4]).mark()))))
+                .toList();
+    }
```
You may want to have in-line comments, it looks like a really complicated return statement to me

zyonwee commented on src/main/resources/images/DaUser.png:
I love this illustration! I am interested in how to cite images

zyonwee commented on src/main/java/fido/task/Deadline.java:
```diff
@@ -0,0 +1,143 @@
+        return new Deadline(super.getName(), false, this.date);
+    }
+
+    /**
+     * Returns a string representation of this {@code Deadline} task in the
+     * format "[D][ ] Task Name (by: MMM dd yyyy)".
+     * - "[D]" denotes it is a {@code Deadline} task.
+     * - "[ ]" or "[X]" indicates whether the task is not done or done respectively.
+     *
+     * @return A formatted {@code String} representing the task.
+     */
```
I like your method comment, it is very neat

zyonwee commented on src/main/resources/view/DialogBox.fxml:
I am interested to know if comments are required for this dialog box, do share with me if you know

zyonwee commented on src/main/java/fido/exception/FidoException.java:
```diff
@@ -0,0 +1,76 @@
+            return "Error: Failed to read from task file.";
+        case FILE_WRITE_ERROR:
+            return "Error: Failed to write to task file.";
+        default:
+            return "Unknown error encountered!";
+        }
```
Perhaps you may want to consider a break case or fall through comment

zyonwee reviewed (commented):
LGTM. Just a few nits to fix.

pang-yy committed:
Use the Streams feature of Java

Almost all codes are already using stream.
So this is just an "empty" commit.

pang-yy committed:
Improve readability of Parser.java

pang-yy committed:
Use assert in Deadline and Event classes

Code in `Deadline.java` and `Event.java` currently does not include any checks or validations
for the number of arguments after splitting the raw input string.

Without validating the number of split arguments,
the program risks encountering index errors or undefined behavior
if the input does not contain the expected number of parts.

Add an `assert` statement in both classes
to ensure the correct number of arguments (2 for Deadline, 3 for Event).

pang-yy committed:
Merge pull request #3 from pang-yy/branch-A-Streams

Use the Streams feature of Java

pang-yy committed:
Merge branch 'master' into branch-A-CodeQuality

pang-yy committed:
Merge pull request #4 from pang-yy/branch-A-CodeQuality

Branch a code quality

pang-yy committed:
Merge branch 'master' into branch-A-Assertions

pang-yy committed:
Merge pull request #5 from pang-yy/branch-A-Assertions

Branch a assertions

pang-yy committed:
Use GitHub Actions to set up Continuous Integration (CI)

pang-yy committed:
Add the ability to recognize and deal with duplicate items

pang-yy committed:
Merge branch 'branch-C-DetectDuplicates'

pang-yy committed:
Add a User Guide to the project

pang-yy committed:
Improve code quality

pang-yy committed:
Fix data file parse error
